neuralpp/experiments/experimental_inference/graph_analysis.py

```diff
@@ -32,39 +33,40 @@ def neighbors(self, node):
 
 class PartialSpanningTree:
 
-    def __init__(self, graph: Graph, root, stop=lambda _: False):
+    def __init__(self, graph: Graph, root):
         self.graph = graph
         self.root = root
         self._children = {}
         self._parents = {}
 
     def children(self, node):
         if not self._children.get(id(node)):
-            neighbors = [n for n in self.graph.neighbors(node) if id(n) not in self._parents]
-            for n in neighbors:
+            available_neighbors = [n for n in self.graph.neighbors(node) if id(n) not in self._parents]
+            for n in available_neighbors:
                 self._parents[id(n)] = node
-            self._children[id(node)] = neighbors
+            self._children[id(node)] = available_neighbors
         return self._children[id(node)]
 
     def parent(self, node):
         return self._parents.get(id(node), None)
 
 
 def node_variables(node):
+    """All variables which are used directly by a node"""
     return node.variables if isinstance(node, Factor) else [node]
 
 
+def variable_at(node):
+    return node if isinstance(node, Variable) else None
+
+
 class FactorPartialSpanningTree(PartialSpanningTree):
 
     def __init__(self, graph: FactorGraph, root):
         super().__init__(graph, root)
-
-        def _evaluate(node):
-            for n in self.children(node):
-                _evaluate(n)
         self._parents[id(root)] = None
-        _evaluate(root)  # Currently evaluates a full spanning tree
 
+    @lru_cache_by_id(1000)
```

Comment:
I am not sure we want a limit here (1000). We might want to store as much as memory will allow to tradeoff memory for speed. If we are going to allow a limit, I think it is better to let that be a parameter of the class (but I am not sure decorators can use those).